### DISCUSSION TITLE
How to format as table?

### DISCUSSION THREAD

**Discussion by Joachimkristensen** [exceljs/exceljs#1672] [Q&A] [answered]
Hi!

I've managed to create a functioning worksheet, but I really want to be able to take my sheet data and make a table out of it. Something similar would be to mark the data and press "Format as Table". 

An other option is to create the table manually, but it seems like the way to create and manage rows/columns is more "approachable" outside a table. I've also made a table using this second approach, but I'm not able to use the same properties on the rows/columns in the table as outside the table (cannot modify column width etc..). 

Do you guys have any tips to approach this issue?
Thanks

**Comment by Siemienik** [comment, marked as answer]
Hi, I fully understand why are you asking :) I had a similar case, and then I create a exceljs template engine for xlsx, which is fully compatible with ExcelJS.

```js
import {Renderer} from 'xlsx-renderer';

const result = await new Renderer().renderFromFile('./report-template.xlsx', reportData);

await result.xlsx.writeFile('./my-awesome-report.xlsx');
```

[Read more about xlsx-renderer](https://github.com/Siemienik/XToolset/tree/master/packages/xlsx-renderer#xlsx-renderer) 

---

When using raw ExcelJS api, the best way is set style for columns, rows, and cells:

`column.style = {font:{bold: true, name: 'Comic Sans MS'}};`

and for rows:

` ws.getRow(2).style = { font: { name: 'Comic Sans MS', family: 4, size: 16, underline: 'double', bold: true }};`

**Reply by Joachimkristensen** [reply]
Thanks for the reply!
I've read about xlsx-renderer and this seems to be a very powerful alternative compared to writing the code for a sheet. However I'm using JSON-data from an external API in my sheets, meaning that some of the columns are going to be dynamic - which works perfectly fine when creating the sheet in JS. Would you still recommend using the xlsx-renderer with this in mind? I'm under the impression that the template sheet is fixed.

**Reply by Siemienik** [reply]
@Joachimkristensen Hmm, It really could be a bottleneck - it depends on what you really need. Thank you for pointing it - I didn't think in that category earlier, which means I can add that feature if needed. 

However, as the xlsx-renderer is built on ExcelJS, it returns the result as ExcelJS's Workbook. So it is possible to clear results from redundant columns and worksheets.

```js
const result = await new Renderer().renderFromFile('./report-template.xlsx', reportData);
result.worksheets[0].spliceColumns(2,1); // remove second col
//...
```

Additionally, the template could be dynamically changed.

```js
const renderer = new Renderer();
const result = await renderer.render(async () => {
            const template = new Workbook();
            const wb = await template.xlsx.readFile(templatePath);
            wb.worksheets[0].spliceColumns(2,1);
            return wb;
        }, viewModel);
//...
```

**Reply by Joachimkristensen** [reply]
I will keep this in mind when working on the template/sheet (whichever I find most compatible with my solution). Appreciate the solid answers. Have a good day :)
